### PR TITLE
Support ByRefLike types as Generic parameters

### PR DESCRIPTION
**Pull request by AaronRobinsonMSFT**
Design - https://github.com/dotnet/runtime/pull/67129

Contributes to https://github.com/dotnet/runtime/issues/65112

Proposed managed API surface for this work - https://github.com/dotnet/runtime/issues/68002

### CODE REVIEW

File: src/coreclr/jit/importer.cpp [hunk 1]
```diff
@@ -4311,6 +4311,32 @@ GenTree* Compiler::impIntrinsic(GenTree*                newobjThis,
                 break;
             }
 
+            case NI_System_Type_get_IsByRefLike:
```
Comment (jkotas) [inline src/coreclr/jit/importer.cpp:4314]: This optimization is (functionally) independent on the byreflike types as generic parameters feature. It can be submitted in separate PR to keep this one smaller.

Comment (AaronRobinsonMSFT) [reply src/coreclr/jit/importer.cpp:4314]: See https://github.com/dotnet/runtime/pull/67852.

File: src/coreclr/vm/i386/jitinterfacex86.cpp [hunk 2]
```diff
@@ -442,6 +442,22 @@ void *JIT_TrialAlloc::GenBox(Flags flags)
     // jne              noAlloc
     sl.X86EmitCondJump(noAlloc, X86CondCode::kJNE);
 
+    // Save the MethodTable ptr
+    sl.X86EmitPushReg(kECX);
```
Comment (jkotas) [inline src/coreclr/vm/i386/jitinterfacex86.cpp:446]: Would it be better to introduce a new JIT box helper and return it from `getBoxHelper` instead of slowing down and complicating the super fast helpers?

Comment (AaronRobinsonMSFT) [reply src/coreclr/vm/i386/jitinterfacex86.cpp]: This was indeed annoying. Lots more changes, but it did help me to find something I missed in NativeAOT.

File: src/coreclr/jit/valuenumfuncs.h [hunk 3]
```diff
@@ -136,6 +136,7 @@ ValueNumFuncDef(JitNewArr, 3, false, true, false)
 ValueNumFuncDef(JitNewMdArr, 4, false, true, false)
 ValueNumFuncDef(JitReadyToRunNew, 2, false, true, false)
 ValueNumFuncDef(JitReadyToRunNewArr, 3, false, true, false)
+ValueNumFuncDef(BoxSlow, 3, false, false, false)
```
Comment (jkotas) [inline src/coreclr/jit/valuenumfuncs.h:139]: This does not need its own value number. It can just use the `Box` value number. (It is similar to how different perf variants of `JIT_New` share same value number.

File: src/coreclr/jit/utils.cpp [hunk 4]
```diff
@@ -1285,7 +1285,6 @@ void HelperCallProperties::init()
             case CORINFO_HELP_NEWSFAST_FINALIZE:
             case CORINFO_HELP_NEWSFAST_ALIGN8_FINALIZE:
             case CORINFO_HELP_READYTORUN_NEW:
```
Comment (jkotas) [inline src/coreclr/jit/utils.cpp:1287]: Should `CORINFO_HELP_BOX` stay here?

Comment (AaronRobinsonMSFT) [reply src/coreclr/jit/utils.cpp:1287]: @jkotas I debated this with someone offline also. I don't think so. This contract states it only throws OOM, which isn't true previously or now. Regardless of what `CORINFO_HELP_BOX` is set to, all implementations eventually call into `JIT_Box` which can throw non-OOM exceptions. My preference here is to ensure the contract is reality so if it can call `JIT_Box`, exceptions are possible. This purity argument could be pushed back against but given the JIT also rarely calls this helper and instead lines it, I am inclined to enforce the purity here.

Comment (jkotas) [reply src/coreclr/jit/utils.cpp:1287]: The (fast) `CORINFO_HELP_BOX` should be only ever called on types for which the slow helper does not throw exceptions (except OOM).

Comment (AaronRobinsonMSFT) [reply src/coreclr/jit/utils.cpp:1287]: Agreed, but that is not what this check this doing. It is describing the nature of the JIT helper, which in this case throws. I agree with your sentiment here, it is identical to my offline conversation. My point though is code is about describing what the JIT helper does not what how the JIT is going to call it.

Comment (AaronRobinsonMSFT) [reply src/coreclr/jit/utils.cpp:1287]: Never mind. I'm not finding my argument as convincing as I once did. I'll move it back and document the expectations.

**Review by jkotas — APPROVED** [review]

File: src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs [hunk 5]
```diff
@@ -2358,12 +2358,13 @@ private CorInfoHelpFunc getBoxHelper(CORINFO_CLASS_STRUCT_* cls)
...
             var type = HandleToObject(cls);
 
-            // we shouldn't allow boxing of types that contains stack pointers
-            // csc and vbc already disallow it.
+            if (type.IsNullable)
+                return CorInfoHelpFunc.CORINFO_HELP_BOX_NULLABLE;
+
             if (type.IsByRefLike)
-                ThrowHelper.ThrowInvalidProgramException(ExceptionStringID.InvalidProgramSpecific, MethodBeingCompiled);
+                return CorInfoHelpFunc.CORINFO_HELP_BOX_SLOW;
```
Comment (MichalStrehovsky) [inline src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs:2365]: Just curious - why do we need a BOX_SLOW helper here? (And a runtime helper that can handle byref-like types at all).

Shouldn't this be a new helper `CORINFO_HELP_BOX_INVALID` that does nothing but throws InvalidProgram at runtime?

Would we ever be in a situation where we need a slow helper that can either box or throw? I would think such helper would only be needed if we have universal shared code (a "`__Canon`" type that could be a byref-like valuetype at runtime). We don't have universal shared code in CoreCLR and __Canon can only be a reference type.

Comment (AaronRobinsonMSFT) [reply src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs:2365]: The BOX_SLOW is the JIT helper that always checks state and basically represents the JIT is naive case. I could add yet another helper for this that just handles all INVALID cases too. I'm not sure that is any better than this approach though. Is the argument that all box failure paths should be detectable at JIT time and we can determine which case then? I've not looked into it, but what about the interpreter?

Comment (MichalStrehovsky) [reply src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs:2365]: From what I saw it looks like we only hand out the SLOW helper when we already know it will throw. If instead if having a SLOW helper we had an INVALID helper that RyuJIT understands is always throwing, it could potentially remove unreachable code after the helper. If it's just a show helper that may or may not throw, it must assume it might return.

Comment (MichalStrehovsky) [reply src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs:2365]: I'm not saying we would do the unreachable optimization now - just as a structural thing.

Comment (AaronRobinsonMSFT) [reply src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs:2365]: @dotnet/jit-contrib Is there prior art in the JIT for indicating a helper will "always throw"? I'm asking because @MichalStrehovsky's suggestion is compelling but adding a new helper implementation that will "always throw" seems a bit premature if there no existing concept. I agree the JIT could use this but unless the JIT is already aware of the concept it seems adding this type of helper now is a bit premature.

Comment (jakobbotsch) [reply src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs:2365]: Yes, there is the `GTF_CALL_M_DOES_NOT_RETURN` flag. Looks like there already are a few helpers we create that we known will throw exceptions using this concept, in particular for unsupported intrinsics. You can use `gtNewMustThrowException` to create a helper call with this flag.

Comment (jakobbotsch) [reply src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs:2365]: It might be better to create a helper `CORINFO_HELP_THROW_INVALIDPROGRAM` and use the above function to create a call to that. There are already other kinds of helpers throwing other exceptions.

Comment (AaronRobinsonMSFT) [reply src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs:2365]: Thanks @jakobbotsch. That makes this decision easy. @MichalStrehovsky I'll update this logic as suggested.

Comment (AaronRobinsonMSFT) [reply src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs:2365]: Well now that I look into this, I am less convinced. Creating either a `CORINFO_HELP_THROW_INVALIDPROGRAM` or `CORINFO_HELP_BOX_INVALID` results in odd statements to determine eliding recognized IL sequences in `Compiler::impBoxPatternMatch`.

```c++
if (boxHelper == CORINFO_HELP_BOX || boxHelper == CORINFO_HELP_THROWINVALIDPROGRAM)
```

or

```c++
if (boxHelper == CORINFO_HELP_BOX || boxHelper == CORINFO_HELP_BOX_INVALID)
```
I personally don't find these very clear and in this case we are now pushing even more policy into the JIT as opposed to letting the VM handle how to fail boxing a ByRefLike. Unless there is a strong case for this potential optimization, which I don't see because the JIT is typically going to inline the box, I am going to leave it as-is.

@MichalStrehovsky @jakobbotsch @jkotas ?

Comment (jakobbotsch) [reply src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs:2365]: I am not super familiar with this handling, @AndyAyersMS might know more. But if I understand correctly `impBoxPatternMatch` tries to optimize common boxes away, and is something we should not be doing in any case if the box helper might throw. If I am right then I would change `impImportBlockCode` to avoid calling that function for `CORINFO_HELP_THROWINVALIDPROGRAM` and fix the else case in `impImportAndPushBox` for the throwing case.

Note that you will need to make sure you spill `exprToBox` to retain side effects of the struct computation, as there are no args to the throwing helper -- the easy way to do that is with `impSpillSideEffects`, but you will need to reorder the code a bit to make sure we do not call `impPopStack` before that.

Comment (AaronRobinsonMSFT) [reply src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs:2365]: > If I am right then I would change impImportBlockCode to avoid calling that function for CORINFO_HELP_THROWINVALIDPROGRAM and fix the else case in impImportAndPushBox for the throwing case.

This is by design for boxing with the ByRefLike generic support - see https://github.com/dotnet/runtime/pull/67129. We need to handle cases where a `box` instruction is created for a `T` that might throw if the branch is taken but isn't in practice.

Comment (jakobbotsch) [reply src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs:2365]: Ah okay, sorry, I have not realized that there are common patterns involving `box` that we still want to support. Then I do think it is a little weird to import a throw helper here and use it to determine this.

Since there is a special contract here for just by-ref like types, would it make sense to just check in the JIT if this is a byref-like type (see e.g. `verIsByRefLike` for a place we do that) and use that to handle the common idioms, and otherwise create the throwing helper? That way we do not involve `getBoxHelper` at all.

Comment (jkotas) [reply src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs:2365]: Maybe you can add a new bool `bool* pInvalidProgram` arg to `getBoxHelper` to allow it to communicate to the JIT that it should throw InvalidProgramException.

Also, `impImportAndPushBox ` is implemented as opportunistic optimization today. It needs revisiting to turn it into a functional guarantee. 

For example, the `box` ; `br_true/false` combination will be optimized only if the tree to box does not have side-effects. vs. https://github.com/dotnet/runtime/pull/67129 says that it will be always optimized and treated as a single unit.

Comment (AaronRobinsonMSFT) [reply src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs:2365]: > Since there is a special contract here for just by-ref like types,

But that is the point of how this was designed, that doesn't exist. The JIT doesn't care what the resolve type is and it defers to the VM to handle it in the best way it sees fit via the slow path.

> Maybe you can add a new bool bool* pInvalidProgram arg to getBoxHelper to allow it to communicate to the JIT that it should throw InvalidProgramException.

I don't see why we should let the JIT make this decision.

> Also, impImportAndPushBox is implemented as opportunistic optimization today. 

Do you mean `impBoxPatternMatch`?

> For example, the box ; br_true/false combination will be optimized only if the tree to box does not have side-effects.

Right. Let me look at that again.

Comment (jakobbotsch) [reply src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs:2365]: > But that is the point of how this was designed, that doesn't exist. The JIT doesn't care what the resolve type is and it defers to the VM to handle it in the best way it sees fit via the slow path.

The JIT is required to optimize several IL patterns for by-ref like types. That is the special contract I am talking about, and why I think this should be checked directly in the JIT. It cannot call the helper provided by the runtime even though it sees a box instruction in these cases, so it does care what the resolve type is.

**Comment by AaronRobinsonMSFT** [comment]
@jakobbotsch and @jkotas Can you please take another look.

Big thanks to @jakobbotsch for the offline help and code examples of what to do in the JIT. Very much appreciated.

File: src/coreclr/inc/corinfo.h [hunk 6]
```diff
@@ -438,7 +438,7 @@ enum CorInfoHelpFunc
     CORINFO_HELP_CHKCASTCLASS_SPECIAL, // Optimized helper for classes. Assumes that the trivial cases
                                     // has been taken care of by the inlined check
 
-    CORINFO_HELP_BOX,
+    CORINFO_HELP_BOX,               // Fast box helper. Only possible exception is OutOfMemory
```
Comment (jkotas) [inline src/coreclr/inc/corinfo.h:441]: This comment is in conflict with the implementation that checks and throws for IsValueType.

Comment (jkotas) [reply src/coreclr/inc/corinfo.h:441]: Delete the check in the implementation and replace it with assert?

Comment (jkotas) [reply src/coreclr/inc/corinfo.h:441]: And/or move it to `CEEInfo::getBoxHelper`

Comment (jkotas) [reply src/coreclr/inc/corinfo.h:441]: Ah ok, I have noticed that you have done it already :-)

Comment (AaronRobinsonMSFT) [reply src/coreclr/inc/corinfo.h:441]: I removed the check for `IsValueType`.

File: src/coreclr/vm/jitinterface.cpp [hunk 7]
```diff
@@ -6013,7 +6013,7 @@ CorInfoHelpFunc CEEInfo::getBoxHelper(CORINFO_CLASS_HANDLE clsHnd)
     }
     else
     {
-        if(VMClsHnd.IsTypeDesc())
+        if (VMClsHnd.IsTypeDesc())
```
Comment (jkotas) [inline src/coreclr/vm/jitinterface.cpp:6016]:
```suggestion
        if (VMClsHnd.IsTypeDesc() || !VMClsHnd.IsValueType())
```

**Comment by jkotas** [comment]
Looks better than before

File: src/coreclr/jit/importer.cpp [hunk 8]
```diff
@@ -7480,7 +7495,7 @@ void Compiler::impImportAndPushBox(CORINFO_RESOLVED_TOKEN* pResolvedToken)
     else
     {
         // Don't optimize, just call the helper and be done with it.
-        JITDUMP(" helper call because: %s\n", canExpandInline ? "optimizing for size" : "nullable");
+        JITDUMP(" helper call because: %s\n", canExpandInline ? "optimizing for size" : "box slow/nullable");
```
Comment (jakobbotsch) [inline src/coreclr/jit/importer.cpp:7498]: Revert this?

File: src/coreclr/jit/compiler.h [hunk 9]
```diff
@@ -4380,10 +4380,17 @@ class Compiler
 
     GenTree* impImportLdvirtftn(GenTree* thisPtr, CORINFO_RESOLVED_TOKEN* pResolvedToken, CORINFO_CALL_INFO* pCallInfo);
 
+    enum BoxPatternFlags
+    {
+        None                  = 0,
+        IsByRefLike           = 1,
+        MakeInlineObservation = 2,
```
Comment (jakobbotsch) [inline src/coreclr/jit/compiler.h:4387]: These names might have a bit high chance to conflict with other names. Prefix with something like `BoxPatterns_`?

File: src/coreclr/jit/importer.cpp [hunk 10]
```diff
@@ -7082,16 +7084,29 @@ int Compiler::impBoxPatternMatch(CORINFO_RESOLVED_TOKEN* pResolvedToken,
...
+                            if (flags & MakeInlineObservation)
                             {
                                 compInlineResult->Note(InlineObservation::CALLEE_FOLDABLE_BOX);
                                 return 1 + sizeof(mdToken);
                             }
 
+                            assert(!(flags & IsByRefLike) || ((impStackTop().val->gtFlags & GTF_SIDE_EFFECT) == 0));
+
                             if (!(impStackTop().val->gtFlags & GTF_SIDE_EFFECT))
                             {
```
Comment (jakobbotsch) [inline src/coreclr/jit/importer.cpp:7096]:
```suggestion
                            if (!(impStackTop().val->gtFlags & GTF_SIDE_EFFECT))
                            {
                                assert((flags & IsByRefLike) == 0);
```
I know this assert was added in my code but I think it is better to just assert inside the if.

Also, now that you're in this function can you update the flag checks to check explicitly against zero? That is the convention we encourage nowadays (https://github.com/dotnet/runtime/blob/main/docs/coding-guidelines/clr-jit-coding-conventions.md#1515-use-of-zero).

File: src/coreclr/jit/importer.cpp [hunk 11]
```diff
@@ -16529,19 +16544,40 @@ void Compiler::impImportBlockCode(BasicBlock* block)
...
                 }
 
-                impImportAndPushBox(&resolvedToken);
-                if (compDonotInline())
+                if (isByRefLike)
                 {
-                    return;
+                    // ByRefLike types are supported in boxing scenarios when the instruction can be elided
+                    // due to a recognized pattern above. If the pattern is not recognized, the code is invalid.
+                    BADCODE("ByRefLike types cannot be boxed");
```
Comment (jakobbotsch) [inline src/coreclr/jit/importer.cpp:16572]: This will throw at JIT time even if the box cannot be reached at runtime. Is that intended?

**Review by jakobbotsch — COMMENTED** [review]
A few nits and one question.

File: src/coreclr/jit/compiler.h (hunk 9 above)
Comment (jkotas) [reply src/coreclr/jit/compiler.h:4387]: Or use `class enum`

File: src/coreclr/jit/importer.cpp (hunk 11 above)
Comment (AaronRobinsonMSFT) [reply src/coreclr/jit/importer.cpp:16572]: Yep. Part of the offline conversation we had was that all we really need to do is ensure the patterns match when a ByRefLike type is involved. Thanks for confirming.

File: src/coreclr/jit/compiler.h (hunk 9 above)
Comment (AaronRobinsonMSFT) [reply src/coreclr/jit/compiler.h:4387]: I guess I could use `class enum` at this point. I was considering these more like flags, but that isn't possible with the current values. I will go with `class enum`.

**Comment by AaronRobinsonMSFT** [comment]
@jakobbotsch Any other feedback here? I'd appreciate a sign-off from the JIT side just so there is awareness on the changes in this code path.

**Review by jakobbotsch — APPROVED** [review]
JIT changes LGTM.

There is `CORINFO_THIS_TRANSFORM::CORINFO_BOX_THIS`. Do we need to do anything special for it or will the VM never return it for byref-like types?

**Comment by AaronRobinsonMSFT** [comment]
> Do we need to do anything special for it or will the VM never return it for byref-like types?

Nothing special. This should just fail naturally in the same way it has always failed since it is not valid for ByRefLike types. This is back to the general solution I originally envisioned vs the narrow approach we are now taking given what we actually need.
